tests/unit/repo/test_graph.py

```diff
@@ -0,0 +1,65 @@
+import pytest
+from networkx import DiGraph
+from networkx.utils import graphs_equal
+
+from dvc.repo.graph import get_steps, get_subgraph_of_nodes
+
+
+@pytest.mark.parametrize(
+    "nodes,downstream,expected_edges",
+    [
+        ([1], False, {1: [2, 3], 2: [4, 5], 3: [6, 7]}),
+        ([2], False, {2: [4, 5]}),
+        ([3], False, {3: [6, 7]}),
+        ([8], False, [(8, 9)]),
+        ([2, 3, 8], False, {2: [4, 5], 3: [6, 7], 8: [9]}),
+        ([4], False, {4: []}),
+        ([1], True, {1: []}),
+        ([9], True, [(8, 9)]),
+        ([2], True, [(1, 2)]),
+        ([6], True, [(1, 3), (3, 6)]),
+        ([2, 3, 8], True, {1: [2, 3], 8: []}),
+        ([4, 7], True, {1: [2, 3], 2: [4], 3: [7]}),
+    ],
+)
+def test_subgraph_of_nodes(nodes, downstream, expected_edges):
+    r"""
+             1
+           /   \
+          2     3      8
+         / \   / \     |
+        4   5 6   7    9
+    """
```

Comment:
Actually, the upstream is downstream, and downstream is upstream for our case, as the graph goes from outs to their dependencies, whereas the steps that we want are from dependencies to outs.

So, imagine here that the graph's direction is from bottom to top, and it'll make sense.
As in if `2` is what you want, the dependencies have to be reproduced which are `[4, 5]`, and then `2` itself. 

The subgraph will be:
```
  2
/  \ 
4   5
```
and the steps will be `[4, 5, 2]`.